### PR TITLE
DataIdentifier

### PR DESCRIPTION
**Pull request by robertbastian**
The new union of `DataLocale` and `DataMarkerAttributes`. Unlike the previous `DataLocale` with auxiliary keys, there's an owned and a borrowed version.

### CODE REVIEW

**Review by Manishearth — DISMISSED** [review]

File: provider/core/src/request.rs [hunk 1]
```diff
@@ -397,26 +536,6 @@ impl DataLocale {
...
-    /// };
-    ///
-    /// let req_with_langid = DataRequest {
-    ///     locale: &langid!("ar-EG").into(),
-    ///     ..Default::default()
-    /// };
-    ///
-    /// assert_eq!(req_no_langid.locale.get_langid(), langid!("und"));
-    /// assert_eq!(req_with_langid.locale.get_langid(), langid!("ar-EG"));
-    /// ```
```
Comment (sffc) [inline provider/core/src/request.rs:419]: Question: why did you delete the test?

File: provider/core/src/request.rs [hunk 2]
```diff
@@ -579,170 +698,94 @@ impl DataLocale {
...
-        } else if let Ok(tiny) = s.parse() {
-            Ok(Self {
-                value: DataMarkerAttributesInner::Stack(tiny),
-            })
-        } else {
-            Ok(Self {
-                value: DataMarkerAttributesInner::Boxed(s.into()),
-            })
-        }
+        Ok(unsafe { core::mem::transmute(s.into_boxed_str()) })
```
Comment (sffc) [inline provider/core/src/request.rs:762]: Issue, here and above: Needs safety comment

Another use case for https://github.com/unicode-org/icu4x/pull/5101 / https://github.com/unicode-org/icu4x/issues/2310 / https://github.com/unicode-org/icu4x/issues/5127

File: provider/datagen/src/driver.rs [hunk 3]
```diff
@@ -847,17 +849,21 @@ fn select_locales_for_marker(
...
-                    if let Some(parent_locales) = maybe_parent_locales {
-                        for morphed_req in parent_locales.iter() {
+                    if let Some(parent_ids) = maybe_parent_ids {
+                        for morphed_id in parent_ids.iter() {
                             // Special case: don't pull extensions or aux keys up from the root.
-                            if morphed_req.0.is_langid_und()
-                                && !(morphed_req.0.is_empty() && morphed_req.1.is_empty())
+                            if morphed_id.locale.is_langid_und()
+                                && !(morphed_id.locale.is_empty()
+                                    && morphed_id.marker_attributes.is_empty())
```
Comment (sffc) [inline provider/datagen/src/driver.rs:857]: Suggestion: can you write this as `!morphed_id.is_empty()` ?

**Review by sffc — CHANGES_REQUESTED** [review]

File: provider/datagen/src/driver.rs (hunk 3 above)
Comment (robertbastian) [reply provider/datagen/src/driver.rs:857]: I don't really want to define too much logic on `DataIdentifier`

**Review by sffc — APPROVED** [review]
Approving to prevent bitrot but there are still open non-blocking comments
